logstash-core/src/main/java/org/logstash/common/io/DeadLetterQueueWriter.java

```diff
@@ -220,6 +255,88 @@ private void innerWriteEntry(DLQEntry entry) throws IOException {
         lastWrite = Instant.now();
     }
 
+    private void executeAgeRetentionPolicy() throws IOException {
+        if (isOldestSegmentExpired()) {
+            deleteExpiredSegments();
+        }
+    }
+
+    private boolean isOldestSegmentExpired() {
+        if (retentionTime == null) {
+            return false;
+        }
+        final Instant now = clock.instant();
+        return oldestSegmentTimestamp
+                .map(t -> t.toInstant().isBefore(now.minus(retentionTime)))
+                .orElse(false);
+    }
+
+    private void deleteExpiredSegments() throws IOException {
+        // remove all the old segments that verifies the age retention condition
+        boolean cleanNextSegment;
+        do {
+            if (oldestSegmentPath.isPresent()) {
+                Path beheadedSegment = oldestSegmentPath.get();
+                deleteTailSegment(beheadedSegment);
+            }
+            updateOldestSegmentReference();
+            cleanNextSegment = isOldestSegmentExpired();
+        } while (cleanNextSegment);
+    }
+
+    private void deleteTailSegment(Path segment) throws IOException {
+        final long segmentSize = Files.size(segment);
+        this.currentQueueSize.add(-segmentSize);
```

Comment:
My first question here was going to ask whether we should check perform the decrement on successful deletion of the file. However, given that `clean_consumed` can remove segments, is the way we are calculating queue sizes still make sense - should we be using it based on actual disk usage instead?

Reply:
Yes, I think that at this point we could move the retrieval of segment files's size outside of the loop where the `deleteTailSegment` is invoked, to avoid too much interaction with the filesystem, in case multiple segments are removed.

Reply:
Fixed with b294d21b25b2467212e90f5e1101bf97953237c1